### PR TITLE
Use QubesDB for knowing if mic is allowed or not

### PR DESCRIPTION
**Pull request by fepitre**
Integration tests: https://github.com/QubesOS/qubes-core-admin/pull/608

### CODE REVIEW

File: qubesguidaemon/mic.py [hunk 1]
```diff
@@ -153,13 +160,40 @@ async def on_device_pre_detach_mic(self, vm, event, device):
...
+                raise qubes.exc.QubesVMError(
+                    vm,
+                    "Failed to detach audio input from {!s} to {!s}: "
+                    "pulseaudio agent not running".format(audiovm, vm),
+                )
+        else:
+            audiovm.untrusted_qdb.write("/audio-input/{}".format(vm.name), "0")
+
+    @qubes.ext.handler("property-set:audiovm")
+    async def on_property_set(
```
Comment (marmarek) [inline qubesguidaemon/mic.py:181]: property set handler can't be async.

**Review by marmarek — CHANGES_REQUESTED** [review]

File: qubesguidaemon/mic.py [hunk 2]
```diff
@@ -153,13 +160,40 @@ async def on_device_pre_detach_mic(self, vm, event, device):
...
+        self, subject, event, name, newvalue, oldvalue=None
+    ):
+        if newvalue == oldvalue:
+            return
+        if oldvalue:
+            mic_allowed = oldvalue.untrusted_qdb.read(
```
Comment (marmarek) [inline qubesguidaemon/mic.py:187]: And another thing found by tests - you can't access qdb when VM is not running (untrusted_qdb is None then).

Comment (marmarek) [reply qubesguidaemon/mic.py:187]: There are three cases:
- "subject" VM is not running - do nothing then
- old audiovm is not running - do nothing (as if mic was not attached)
- new audiovm is not running - here I'm not sure, especially if mic was attached; at least log a warning

File: qubesguidaemon/mic.py [hunk 3]
```diff
@@ -153,13 +160,40 @@ async def on_device_pre_detach_mic(self, vm, event, device):
...
+                "/audio-input/{}".format(subject.name)
             )
-        except subprocess.CalledProcessError:
-            raise qubes.exc.QubesVMError(
-                vm,
-                "Failed to detach audio input from {!s} to {!s}: "
-                "pulseaudio agent not running".format(audiovm, vm),
+            if mic_allowed not in (b"0", b"1"):
```
Comment (marmarek) [inline qubesguidaemon/mic.py:190]: This can be also missing (`None`? needs checking), when relevant pacat process is not running (or is an older version)

**Comment by marmarek** [comment]
> * TC_20_AudioVM_Pulse_debian-12-xfce: [test_252_audio_playback_audiovm_switch_hvm](https://openqa.qubes-os.org/tests/105938#step/TC_20_AudioVM_Pulse_debian-12-xfce/7) (error)
>       `qubes.exc.QubesVMError: Invalid value for /audio-input/test-inst-vm...`

Unfortunately error message doesn't tell what value was there actually...

> * TC_20_AudioVM_Pulse_fedora-40-xfce: [test_222_audio_rec_unmuted_pulseaudio](https://openqa.qubes-os.org/tests/105939#step/TC_20_AudioVM_Pulse_fedora-40-xfce/3) (error)
>       `qubes.devices.DeviceAlreadyAttached: 'device dom0:mic of class mic ...`

First test passed and only later one failed. I think it's about stale entry from previous VM startup. Remove it in `domain-qdb-create` event handler (it's part of domain startup).

**Comment by marmarek** [comment]
> * TC_20_AudioVM_PipeWire_fedora-40-xfce: [test_250_audio_playback_audiovm_pipewire](https://openqa.qubes-os.org/tests/105938#step/TC_20_AudioVM_PipeWire_fedora-40-xfce/4) (error)
>       `AttributeError: 'NoneType' object has no attribute 'is_running'`

And also this issue

**Comment by marmarek** [comment]
> * TC_20_AudioVM_PipeWire_fedora-40-xfce: [test_260_audio_mic_enabled_switch_audiovm](https://openqa.qubes-os.org/tests/105938#step/TC_20_AudioVM_PipeWire_fedora-40-xfce/6) (failure)
>       `AssertionError: source-output for VM test-inst-vm1 not found`

This is the actual test. My guess is `_configure_audio_recording` doesn't handle recording with audiovm, it looks to call pactl only in dom0. `common_audio_record_unmuted` looks to also support only recording with dom0 as audiovm... It might be enough to prepend `qvm-run -p ...` to those commands in case of audiovm.

**Comment by fepitre** [comment]
> > * TC_20_AudioVM_Pulse_debian-12-xfce: [test_252_audio_playback_audiovm_switch_hvm](https://openqa.qubes-os.org/tests/105938#step/TC_20_AudioVM_Pulse_debian-12-xfce/7) (error)
> >   `qubes.exc.QubesVMError: Invalid value for /audio-input/test-inst-vm...`
> 
> Unfortunately error message doesn't tell what value was there actually...

Yes, I thought it would be a bad idea to display it, I will adapt it.

File: qubesguidaemon/mic.py [hunk 4]
```diff
@@ -153,13 +160,53 @@ async def on_device_pre_detach_mic(self, vm, event, device):
...
+            mic_allowed = oldvalue.untrusted_qdb.read(
+                "/audio-input/{}".format(subject.name)
             )
-        except subprocess.CalledProcessError:
-            raise qubes.exc.QubesVMError(
-                vm,
-                "Failed to detach audio input from {!s} to {!s}: "
-                "pulseaudio agent not running".format(audiovm, vm),
+            mic_allowed_value = (
+                mic_allowed.decode("utf-8")
```
Comment (marmarek) [inline qubesguidaemon/mic.py:196]: since you care about `"0"` and `"1"` only, better use `ascii`; and in any case, handle UnicodeDecodeError

Comment (marmarek) [reply qubesguidaemon/mic.py:196]: or simply compare against `b"0"` and `b"1"`

File: pulse/pacat-simple-vchan.c [hunk 5]
```diff
@@ -956,10 +952,8 @@ static void control_cleanup(struct userdata *u) {
         u->mainloop_api->io_free(u->control_socket_event);
     if (u->control_socket_fd > 0)
         close(u->control_socket_fd);
-    if (u->qdb && u->qdb_path)
-        qdb_rm(u->qdb, u->qdb_path);
-    if (u->qdb_path)
-        free(u->qdb_path);
+    if (u->qdb_config_path)
+        free(u->qdb_config_path);
```
Comment (marmarek) [inline pulse/pacat-simple-vchan.c:956]: free other paths too?

and since we have a separate config and status paths, maybe restore removing status path (to signal it isn't really connected anymore)?

File: pulse/pacat-simple-vchan.c [hunk 6]
```diff
@@ -935,9 +931,9 @@ static int setup_control(struct userdata *u) {
     return 0;
 
 fail:
-    if (u->qdb_path)
-        free(u->qdb_path);
-    u->qdb_path = NULL;
+    if (u->qdb_config_path)
+        free(u->qdb_config_path);
+    u->qdb_config_path = NULL;
```
Comment (marmarek) [inline pulse/pacat-simple-vchan.c:936]: free other paths on failure too

**Review by marmarek — CHANGES_REQUESTED** [review]

**Comment by marmarek** [comment]
https://openqa.qubes-os.org/tests/107154#step/update_guivm/22
```
Jul 21 13:05:34 dom0 qubesd[1412]: unhandled exception while calling src=b'dom0' meth=b'admin.vm.Start' dest=b'sys-gui-vnc' arg=b'' len(untrusted_payload)=0
Jul 21 13:05:34 dom0 qubesd[1412]: Traceback (most recent call last):
Jul 21 13:05:34 dom0 qubesd[1412]:   File "/usr/lib/python3.11/site-packages/qubes/api/__init__.py", line 291, in respond
Jul 21 13:05:34 dom0 qubesd[1412]:     response = await self.mgmt.execute(
Jul 21 13:05:34 dom0 qubesd[1412]:                ^^^^^^^^^^^^^^^^^^^^^^^^
Jul 21 13:05:34 dom0 qubesd[1412]:   File "/usr/lib/python3.11/site-packages/qubes/api/admin.py", line 866, in vm_start
Jul 21 13:05:34 dom0 qubesd[1412]:     await self.dest.start()
Jul 21 13:05:34 dom0 qubesd[1412]:   File "/usr/lib/python3.11/site-packages/qubes/vm/qubesvm.py", line 1277, in start
Jul 21 13:05:34 dom0 qubesd[1412]:     self.create_qdb_entries()
Jul 21 13:05:34 dom0 qubesd[1412]:   File "/usr/lib/python3.11/site-packages/qubes/vm/qubesvm.py", line 2397, in create_qdb_entries
Jul 21 13:05:34 dom0 qubesd[1412]:     self.fire_event('domain-qdb-create')
Jul 21 13:05:34 dom0 qubesd[1412]:   File "/usr/lib/python3.11/site-packages/qubes/events.py", line 195, in fire_event
Jul 21 13:05:34 dom0 qubesd[1412]:     sync_effects, async_effects = self._fire_event(event, kwargs,
Jul 21 13:05:34 dom0 qubesd[1412]:                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jul 21 13:05:34 dom0 qubesd[1412]:   File "/usr/lib/python3.11/site-packages/qubes/events.py", line 164, in _fire_event
Jul 21 13:05:34 dom0 qubesd[1412]:     effect = func(self, event, **kwargs)
Jul 21 13:05:34 dom0 qubesd[1412]:              ^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jul 21 13:05:34 dom0 qubesd[1412]:   File "/usr/lib/python3.11/site-packages/qubesguidaemon/mic.py", line 226, in on_domain_qdb_create
Jul 21 13:05:34 dom0 qubesd[1412]:     vm.audiovm.untrusted_qdb.rm("/audio-input-config/{}".format(vm.name))
Jul 21 13:05:34 dom0 qubesd[1412]:     ^^^^^^^^^^^^^^^^^^^^^^^^
Jul 21 13:05:34 dom0 qubesd[1412]: AttributeError: 'NoneType' object has no attribute 'untrusted_qdb'
```

Since it accesses `vm.audiovm.untrusted_qdb`, it needs to check if it's running, similar to other places.

**Review by marmarek — APPROVED** [review]
